### ISSUE TITLE
Rate not being set by custom profile

### ISSUE THREAD

**Issue by Nate-Wessel** [Project-OSRM/osrm-backend#4837]
I've written a very simple custom profile in an effort to understand how profiles work before getting into the weeds and building something more complicated for my application. I've found however that the rates that I set explicitly in the profile are not being stored in the graph. I can set the speed and the name of an edge and see these values returned in the debug vector tiles. However no matter what I do, it does not seem to accept the custom rate values that I've been trying to supply, but rather calculating these based on speed. Shouldn't an explicit value override a default?

My profile is below. The gist of it is that I'm trying to give preference to known transit routes. 

````
-- TRANSIT OSRM profile for retro-gtfs project
api_version = 4

Set = require('lib/set')
Sequence = require('lib/sequence')
Handlers = require("lib/way_handlers")
Relations = require("lib/relations")
find_access_tag = require("lib/access").find_access_tag
limit = require("lib/maxspeed").limit
Utils = require("lib/utils")

-- set properties
function setup()
	return {
		properties = {
			weight_name = 'custom'
		},
		-- check these tags for access restrictions, in this order
		access_tags_to_check = {
			'ttc','psv','bus','motor_vehicle','vehicle','access'
		},
		-- 
		access_values_to_accept = Set {
			'yes','designated','destination','permissive'
		},
		-- WHY IS THIS NECESSARY?
		relation_types = Sequence {
			"route"
		}
	}
end

-- process all nodes, really just looking for barriers here
function process_node(profile, node, result, relations)
	-- is this node a barrier?
	local barrier = node:get_value_by_key("barrier")
	if barrier then
		for i, access_tag in ipairs(profile.access_tags_to_check) do
			-- does this access tag exist and have an acceptable value?
			local access_value = node:get_value_by_key(access_tag)
			if access_value and profile.access_values_to_accept[access_value] then 
				-- not a barrier to transit
				print('barrier:',barrier,'access_tag',access_tag,'=',access_value)
				return
			end -- if acceptable access value
		end -- for access tags to check
		-- there was a barrier but we havn't cleared it
		result.barrier = true
	end -- if barrier
end


function process_way(profile, way, result, relations)
	railway = way:get_value_by_key('railway')
	highway = way:get_value_by_key('highway')
	route = way:get_value_by_key('route')
	-- perform a quick initial check and abort if the way is
	-- obviously not routable.
	-- highway or route or railway tags must exist
	if (not highway) and (not route) and (not railway) then
		return
	end
	-- now check for any access tags
	for i, access_tag in ipairs(profile.access_tags_to_check) do
		-- does this access tag exist and have an acceptable value?
		local access_value = way:get_value_by_key(access_tag)
		if access_value and not profile.access_values_to_accept[access_value] then 
			-- not a barrier to transit
			print('denied way with access_tag',access_tag,'=',access_value)
			return 
		end -- if acceptable access value
	end -- for access tags to check
	-- set the default travel mode
	result.forward_mode = mode.driving
	result.backward_mode = mode.driving
	-- set default speeds
	result.forward_speed = 20
	result.backward_speed = 20
	-- set default rates
	result.forward_rate = 5.0
	result.backward_rate = 5.0
	-- set a name
	result.name = way:get_value_by_key('name')
	-- HANDLE ANY RELATION DATA
	-- get a list of relations
	local rel_id_list = relations:get_relations(way)
	-- iterate over them
	for i, rel_id in ipairs(rel_id_list) do
		-- get the relation object
		local rel = relations:relation(rel_id)
		-- find the type of relation
		local reltype = rel:get_value_by_key("type")
		-- is the relation a route?relation
		if reltype == 'route' then
			-- is the route a transit route?
			local route = rel:get_value_by_key("route")
			local route_name = rel:get_value_by_key("name")
			if route and ( route == 'bus' or route == 'tram' ) then
				-- let's be sure we've done something
				local street_name = way:get_value_by_key("name")
				print('route',route_name,'on',street_name)
				-- adjust the weight and rate
				result.forward_speed = 40
				result.backward_speed = 40
				result.forward_rate = 10.0
				result.backward_rate = 10.0
				return result
			end -- if transit route
		end -- if route
	end -- for relation
	return result
end -- way function

function process_turn(profile, turn)
--[[
nothing here for now
--]]
end

return {
  setup = setup,
  process_way = process_way,
  process_node = process_node,
  process_turn = process_turn
}
````

Where I set the speed to 20, the edges are given a rate of 0.5. Where they are set to 40, I see rates of 1.0. What is going wrong here? The rate is being set in the same part of the code as the speed, it's just not being accepted.

**Comment by Nate-Wessel** [comment]
I've spent a few days looking at the [profile documentation page](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/profiles.md) now, and I think there may be an important distinction not being made. Some features appear to be implemented only in the three provided Lua profile scripts, while others are features of the C backend. `weight_name` for example has ambiguous or no effects on routing outcomes (that I can see) when used in a custom profile. Is that because it's used as a switch for various other settings in one particular profile? In any case, the effect on a custom profile is undocumented. Another example is in the description of the `process_node` function. It says you can set `result.traffic_light = true` and that this is used in the `process_turn` function to actually set a delay value. In a custom profile there may be no such handling. I would suggest that documentation on profiles as such be on one page and documentation on how people can make minor modifications to the provided profiles on another. 

It might also help to provide a bare-bones sample.lua profile, with minimal dependency on other files that could provide a template for people to make their own profiles, without needing to read through all the way-handler etc files.

**Comment by TheMarex** [comment]
To debug this I would recommend adding some debugging code [here](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/extractor_callbacks.cpp#L395) that prints `forward_weight_data(100)` and `backward_weight_data(100)`. If you set a rate of `5`, the expected result would be `100 / 5 = 20`.

**Comment by Nate-Wessel** [comment]
I just tried setting a few different rates in my profile:
When the rate is set to 5 the code you suggested did output 20, and it output 10 when the rate was set to 10. 
I see in the tiles that the rate now shows as 0.5 where it was set to 5 and 1 where it was set to 10. If I set it to 20, I get 2. So perhaps this is just a misplaced decimal issue? I've also noticed that if I change the weight_precision property to anything > 1 the reported rate goes down to 0 across the board while the value of the weights increases. The values returned for rate in the tiles seem to only ever have one decimal place.
